### ISSUE TITLE
How can I plot feature representation into 2D spaces? as in figures?

### ISSUE THREAD

**Issue by LifeBeyondExpectations** [YangHM/Convolutional-Prototype-Learning#2]
Does paper use t-sne or PCA ?? ..

**Comment by WeihongM** [comment]
@LifeBeyondExpectations    @YangHM 
same problem?
Do you know how to plot now?

**Comment by fjchange** [comment]
A bottleneck layers with only two units may solve it. The feature will plot into 2-D space, then we can visualize them. 
In nets.py, the last but one layers is the bottleneck layers. It is  commonly used in toy model to visualize the ability of a new loss function.

**Comment by TLMichael** [comment]
@fjchange But I just tried to train the model with a bottleneck layer(2 dimensions) on MNIST by softmax+cross entropy loss, it seems can't converge. The accuracy keeps at ~11.35.

**Comment by YangHM** [comment]
> A bottleneck layers with only two units may solve it. The feature will plot into 2-D space, then we can visualize them.
> In nets.py, the last but one layers is the bottleneck layers. It is commonly used in toy model to visualize the ability of a new loss function.

You are right. The feature layer only has two nodes, thus the extracted features by the CNN are two dimensional. For the two dimensional features, we can directly plot them for visualization. I found this method on other works, so l follow them to conduct the same toy experiments.

**Comment by YangHM** [comment]
> @fjchange But I just tried to train the model with a bottleneck layer(2 dimensions) on MNIST by softmax+cross entropy loss, it seems can't converge. The accuracy keeps at ~11.35.

I don't know the correct definition of the "bottleneck layer". In our work, we directly set the feature layer (second to last layer) with only two nodes. For the convergence problem, there may be different reasons.    One main reason is the setting of the hyper-parameters, the other reason is the data used, we don't use the raw data but with some pre-processing. Please read the README.md of this repository for more information about the data.

**Comment by TLMichael** [comment]
> > @fjchange But I just tried to train the model with a bottleneck layer(2 dimensions) on MNIST by softmax+cross entropy loss, it seems can't converge. The accuracy keeps at ~11.35.
> 
> I don't know the correct definition of the "bottleneck layer". In our work, we directly set the feature layer (second to last layer) with only two nodes. For the convergence problem, there may be different reasons. One main reason is the setting of the hyper-parameters, the other reason is the data used, we don't use the raw data but with some pre-processing. Please read the README.md of this repository for more information about the data.

Then it makes sense since I used raw MNIST dataset without pre-processing. Thank you!

**Comment by yangrunz** [comment]
Do you have the program to plot  the 2-D feature space of the sample? I have no ideal about how to draw it

**Comment by WeihongM** [comment]
@yangrunz  in my opinion, you just need to use addtional conv layer, which output dim is shape2.
And use the output to plot 2-D feature

**Comment by yangrunz** [comment]
ok，l will take a try .Thanks for your reply.



发自我的iPhone


------------------ Original ------------------

**Comment by jiejiangd** [comment]
> > > @fjchange But I just tried to train the model with a bottleneck layer(2 dimensions) on MNIST by softmax+cross entropy loss, it seems can't converge. The accuracy keeps at ~11.35.
> > 
> > 
> > I don't know the correct definition of the "bottleneck layer". In our work, we directly set the feature layer (second to last layer) with only two nodes. For the convergence problem, there may be different reasons. One main reason is the setting of the hyper-parameters, the other reason is the data used, we don't use the raw data but with some pre-processing. Please read the README.md of this repository for more information about the data.
> 
> Then it makes sense since I used raw MNIST dataset without pre-processing. Thank you!

Hello, do you have the author's preprocessed MNIST data? The author said that we could send an email to ask for it, but I haven't got a reply. Thank you!
